File: server_py/validation.py
class ApiError(Exception):
    def __init__(self, message, status=400):
        super().__init__(message)
        self.status = status


def clean_rut(rut):
    return str(rut or "").replace(".", "").replace(" ", "").upper()


def valid_rut(rut):
    value = clean_rut(rut)
    if "-" not in value:
        return False
    body, verifier = value.split("-", 1)
    if not body.isdigit() or len(body) not in (7, 8) or verifier not in "0123456789K":
        return False
    total = 0
    multiplier = 2
    for digit in reversed(body):
        total += int(digit) * multiplier
        multiplier = 2 if multiplier == 7 else multiplier + 1
    expected_value = 11 - (total % 11)
    expected = "0" if expected_value == 11 else "K" if expected_value == 10 else str(expected_value)
    return verifier == expected


def public_user(user):
    if not user:
        return None
    result = dict(user)
    result.pop("clave", None)
    return result

File: server/validation.js
export function cleanRut(rut) {
  return String(rut || '').replace(/\./g, '').replace(/\s/g, '').toUpperCase()
}

export function isValidRut(rut) {
  const value = cleanRut(rut)
  if (!/^\d{7,8}-[\dK]$/.test(value)) return false
  const [body, verifier] = value.split('-')
  let sum = 0
  let multiplier = 2
  for (let index = body.length - 1; index >= 0; index -= 1) {
    sum += Number(body[index]) * multiplier
    multiplier = multiplier === 7 ? 2 : multiplier + 1
  }
  const expectedValue = 11 - (sum % 11)
  const expected = expectedValue === 11 ? '0' : expectedValue === 10 ? 'K' : String(expectedValue)
  return verifier === expected
}

export function publicUser(user) {
  if (!user) return null
  const { clave: _clave, ...safeUser } = user
  return safeUser
}

export function normalizeRole(role) {
  const value = String(role || '').toLowerCase()
  return ['gerente', 'mesero', 'cocinero'].includes(value) ? value : ''
}

export function badRequest(message) {
  const error = new Error(message)
  error.status = 400
  return error
}

export function forbidden(message) {
  const error = new Error(message)
  error.status = 403
  return error
}

export function notFound(message) {
  const error = new Error(message)
  error.status = 404
  return error
}
